Dockerfile

```diff
@@ -1,22 +1,41 @@
 # Base will install runtime dependencies and configure generics
-FROM node:12.16
+FROM node:12-slim as base
+LABEL maintainer="Marko Kajzer <markokajzer91@gmail.com>, Nico Stapelbroek <discord-soundbot@nstapelbroek.com>"
+
+RUN apt-get update -q && apt-get install -qy wget
+RUN wget -qO /tini https://github.com/krallin/tini/releases/download/v0.18.0/tini-$(dpkg --print-architecture) && \
+    chmod +x /tini && mkdir /app && chown -R node:node /app
 
-LABEL maintainer="Marko Kajzer <markokajzer91@gmail.com>"
 WORKDIR /app
 
-# Copy some files
-COPY package.json yarn.lock ./
-COPY . /app
 
-# Install ffmpeg and other deps
-RUN apt-get update
-RUN apt-get install --yes build-essential git ffmpeg make python > /dev/null
-RUN yarn install --silent
+# Builder will install system dependencies
+FROM base as builder
+RUN apt-get update -q && apt-get install -qy python make g++ tar xz-utils
+RUN wget -qO /tmp/ffmpeg.tar.xz https://johnvansickle.com/ffmpeg/releases/ffmpeg-release-$(dpkg --print-architecture)-static.tar.xz && \
+    tar -x -C /usr/local/bin --strip-components 1 -f /tmp/ffmpeg.tar.xz --wildcards '*/ffmpeg' && rm /tmp/ffmpeg.tar.xz
+
 
-# Build
+# Build will compile ts to js
+FROM builder AS build
+USER node
+COPY --chown=node:node package.json yarn.lock ./
+RUN yarn install --frozen-lockfile --ignore-optional
+
+COPY . /app
 RUN yarn build
+RUN yarn install --frozen-lockfile --production --ignore-optional && yarn cache clean --force
+
 
-# Cleanup
-RUN apt-get remove --yes build-essential > /dev/null
+# release has the bare minimum to run the application
+FROM base as release
+ENV NODE_ENV=production
+COPY --chown=node:node --from=build /usr/local/bin/ffmpeg /usr/local/bin/ffmpeg
+COPY --chown=node:node --from=build /app/config ./config
+COPY --chown=node:node --from=build /app/sounds ./sounds
+COPY --chown=node:node --from=build /app/node_modules ./node_modules
+COPY --chown=node:node --from=build /app/dist ./dist
 
-CMD ["yarn", "serve"]
+USER node
+ENTRYPOINT ["/tini", "--"]
+CMD ["node", "./dist/bin/soundbot.js"]
```

Comment:
```suggestion
CMD ["yarn", "serve"]

```

Reply:
Out of curiosity, what does this do? 😮 Is there some path rewrite or magic that happens when running via npm/yarn?

Reply:
It's the same to be honest. `yarn serve` does `node .dist/bin/soundbot.js`

Since yarn is installed we can use it, and if I change the file structure, `yarn serve` will probably be kept up to date, while I'll easily forget about the Dockerfile. 🤔